We calculated the phenotypic variance explained by genotypes by using GCTA (v 1.26.0) (https://cnsgenomics.com/software/gcta/#Overview) with the following steps:

Step 1. Estimate the genetic relationship matrix (GRM) between pairs of individuals from SNPs or SVs and save the lower triangle elements of the GRM to binary files. Take SNPs data for an examples. Note that ./SNPs_106.ped, ./SNPs_106.bim, and ./SNPs_106.fam files were previously generated by plink under the ./ folder. 
```
$GCTA_FOLDER/gcta64 --bfile ./SNPs_106 --autosome-num 10 --autosome --make-grm --out ./SNPs_106
```

Step 2. Perform a REML (restricted maximum likelihood) analysis and estimate variance explained by all the SNPs. The ./phenotype.SNP.trait is a file with 3 columns: family ID, individual ID, phenotype (numeric). 
```
$GCTA_FOLDER/gcta64 --grm ./SNPs_106 --pheno ./phenotype.SNP.trait --reml --reml-no-constrain --out ./phenotype.SNP.trait.all_common_snps
```
