We first downloaded 11 geomes from the link: 

Then we use the following steps to generate the SV set: 

First, download these genomes fasta file at https://download.maizegdb.org/:
Zm-B73-REFERENCE-GRAMENE-4.0.fa
Zm-A188-REFERENCE-KSU-1.0.fa
Zm-F7-REFERENCE-TUM-1.0.fa
Zm-EP1-REFERENCE-TUM-1.0.fa
Zm-K0326Y-REFERENCE-SIPPE-1.0.fa
Zm-PH207-REFERENCE_NS-UIUC_UMN-1.0.fa
Zm-W22-REFERENCE-NRGENE-2.0.fa
Zm-PE0075-REFERENCE-TUM-1.0.fa
Zm-CML247-REFERENCE-PANZEA-1.1.fa
Zm-B104-DRAFT-ISU_USDA-0.1.fa
Zm-DK105-REFERENCE-TUM-1.0.fa
Zm-NC358-REFERENCE-NAM-1.0.fa


Second, convert scaffolds into contigs,(to remove multi Ns gap in sequence).
perl convert_scaffolds_into_contigs.pl genome.fa genome.contig.fa (genome.fa could be one of genome file above)

Third, run nucmer (mummer4.00 https://github.com/mummer4/mummer) with commands below:
nucmer -t 60 -c 90 -l 40 -p	B97	Zm-Mo17-REFERENCE-CAU-1.0.fa	genome.contig.fa

Fourth, genome.delta files generated by nucmer were upload to http://assemblytics.com/, to detect SV(Unique sequence length required = 10000;Maximum variant size = 10000;Minimum variant size = 50).
Output files of assemblytics is in .bed format. Use SURVIVOR convertAssemblytics https://github.com/fritzsedlazeck/SURVIVOR to convert .bed files into .vcf files.
Then use SURVIVOR merge to combine 12 .vcf files.
SURVIVOR merge vcflist 1000 1 1 -1 -1 -1 12_genome.intersect.1000.vcf
